source/common/router/router.cc

```diff
@@ -633,12 +635,15 @@ Filter::HttpOrTcpPool Filter::createConnPool(Upstream::HostDescriptionConstShare
 }
 
 UpstreamRequestPtr Filter::createUpstreamRequest(Filter::HttpOrTcpPool conn_pool) {
+  GenericConnPoolFactory* factory;
   if (absl::holds_alternative<Http::ConnectionPool::Instance*>(conn_pool)) {
-    return std::make_unique<UpstreamRequest>(*this,
-                                             std::make_unique<HttpConnPool>(*httpPool(conn_pool)));
+    factory = &Envoy::Config::Utility::getAndCheckFactoryByName<GenericConnPoolFactory>(
```

Comment:
It might be worth sketching out what this would look like from a config perspective? Have you thought about that? It's not completely clear to me how the user might select an override conn pool and what variables we want that selection to take place on. cc @lambdai for more input.

Additionally, as I mentioned, there is a general problem I would really like to solve with this extension point which is to have the ability for Envoy to use centralized state for load balancing. Let's say for example that someone wanted to write an extension where we talk to redis to fetch an RR index or a host to load balance to. I *think* that with this extension point we could do that. For example, we could have an extension "CentralizedStateConnPoolFactory" which would do something like:
1) Talk to a centralized cache/DB to select a host.
2) Create a conn pool targeting that host.

Does this sound reasonable as something that could be done with this extension point?

Reply:
Right now, with TODO/FIXME, the router selects an HttpOrTcpConnPool and hands it to the createGenericConnPool call, which creates a http wrapper or a tcp wraper around the HttpOrTcpPool.

I think what we want  to do is move the router's createConnPool to the factory, so it can on its own select config_.cm_.tcpConnPoolForCluster() or config_.cm_.httpConnPoolForCluster() or config_cm.myWeirdCustomConnPoolForCluster()

I'd have an easier time reasoning about it if I did the move first, then addressed the config as I address that TODO, but if you think it's easier to reason about the other way I can drop this for now and try to move the xyzConnPoolForCluster calls to the "pluggable" Tcp/Http conn pools with config first?

Reply:
My preference would be to try to figure out the factory first before doing other work, but up to you.